src/Umbraco.Infrastructure/Persistence/Repositories/Implement/SimilarNodeName.cs

```diff
@@ -1,118 +1,249 @@
 
-                if (xpos < 0 || ypos < 0 || xpos != ypos)
-                    return string.Compare(xname, yname, StringComparison.Ordinal);
+                return model.FullName;
+            }
+
+            // has suffix - name without suffix also exists, therefore we simply increment
+            if (model.HasSuffix() && items.SimpleNameExists(model.Text))
+            {
+                var nextSuffix = GetSuffixNumber(items);
+                model.SetSuffix(nextSuffix);
 
-                // compare the part before (number)
-                var n = string.Compare(xname, 0, yname, 0, xpos, StringComparison.Ordinal);
-                if (n != 0)
-                    return n;
+                return model.FullName;
+            }
 
-                // compare (number) lengths
-                var diff = xname.Length - yname.Length;
-                if (diff != 0) return diff < 0 ? -1 : +1;
+            return name;
+        }
 
-                // actually compare (number)
-                var i = xpos;
-                while (i < xname.Length - 1)
+        private static int GetFirstSuffix(IEnumerable<StructuredName> items)
+        {
+            const int suffixStart = 1;
+
+            if (!items.Any(x => x.Number == suffixStart))
+            {
+                // none of the suffixes are the same as suffixStart, so we can use suffixStart!
+                return suffixStart;
+            }
+
+            return GetSuffixNumber(items);
+        }
+
+        private static int GetSuffixNumber(IEnumerable<StructuredName> items)
+        {
+            int current = 1;
+            foreach (var item in items.OrderBy(x => x.Number))
+            {
+                if (item.Number == current)
                 {
-                    if (xname[i] != yname[i])
-                        return xname[i] < yname[i] ? -1 : +1;
-                    i++;
+                    current++;
+                }
+                else if (item.Number > current)
+                {
+                    // do nothing - we found our number!
+                    // eg. when suffixes are 1 & 3, then this method is required to generate 2
+                    break;
                 }
-                return 0;
             }
+
+            return current;
         }
 
-        // gets a unique name
-        public static string GetUniqueName(IEnumerable<SimilarNodeName> names, int nodeId, string nodeName)
+        internal class StructuredName
         {
-            var uniqueNumber = 1;
-            var uniqueing = false;
-            foreach (var name in names.OrderBy(x => x, new Comparer()))
+            private const string Suffixed_Pattern = @"(.*)\s\((\d+)\)$";
+            internal string Text { get; private set; }
+            internal int Number { get; private set; }
+            public string FullName
+            {
+                get
+                {
+                    string text = string.IsNullOrWhiteSpace(Text) ? Text.Trim() : Text;
```

Comment:
If `Text` is `null` then Trim throws? Should this not have been the opposite?

Reply:
In this case I should not have used IsNullOrWhiteSpace. Due to the replacement with Space_Character constant, Text will always at least have a value of a space character. It should only need to test for whitepsace. 

So it should be something like this...
var text = Text = Space_Character : Text.Trim() : Text

Reply:
`string text = (Text == Space_Character) ? Text.Trim() : Text;` make sense, yes :)